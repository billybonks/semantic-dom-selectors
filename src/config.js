class Config {
  constructor() {
    this.reset();
  }

  reset() {
    this.registeredFinders = [];
    this.defaultFinders = [];
    this.errorLevelOptions = {};
    this.customActors = {
      select: [],
      text: [],
      toggle: [],
      button: [],
    };
    this._rootElement = document.documentElement;
  }

  trim(text) {
    return text.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
  }


  get rootElement() {
    return this._rootElement;
  }

  set rootElement(element) {
    if (element instanceof Element) {
      this._rootElement = element;
    } else {
      throw new Error(`rootElement Error: You tried to set root element to ${element}`);
    }
  }

  registerFinder(finder) {
    if (!this.errorLevelOptions[finder.key]) {
      this.errorLevelOptions[finder.key] = 1;
    }
    this.registeredFinders.push(finder);
  }

  registerActor({ type, run }) {
    if (!type) {
      throw new Error('You must specify the type of attribute of actor object');
    }
    if (!Object.keys(this.customActors).includes(type)) {
      throw new Error(`Unkown actor type ${type}`);
    }
    this.customActors[type].push(run);
  }

  setErrorLevels(config) {
    this.errorLevelOptions = config;
  }

  get actors() {
    return this.customActors;
  }

  get finders() {
    return this.registeredFinders.concat(this.defaultFinders);
  }

  get rules() {
    const rules = this.finders.reduce((acc, finder) => {
      const config = this.errorLevelOptions[finder.key];
      if (isNaN(config)) {
        acc[finder.key] = 1;
      } else {
        acc[finder.key] = config;
      }
      return acc;
    }, {});
    return rules;
  }
}

export default new Config();
