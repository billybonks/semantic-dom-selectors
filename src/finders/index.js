import findByAria from './find-by-aria/index';
import findByLabel from './find-by-label';
import findByName from './find-by-name';
