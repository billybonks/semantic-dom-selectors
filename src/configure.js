import config from './config';
import mergePresets from './config/merge-presets';

/**
 * Configures the library to behave in a certain way
 *
 *
 * @param {object} configuration
 *
 * @example
 * let config = {
 *  preset: 'aria-compliant',
 *  rules: {
 *    'something-else': 1,
 *    'invalid-label-for': 2,
 *  },
 *  finders: [
 *    { key: 'asd', run() {} },
 *    { key: 'qqq-www', run() {} },
 *  ],
 *  actors: [
 *    { key: 'some-act', type: 'select', run() {} },
 *    { key: 'some-input-act', type: 'input', run() {} },
 *  ],
 * };
 *
 * configure(config)
 */
export default function (configuration) {
  const normalizedConfig = mergePresets(configuration);
  config.reset();
  normalizedConfig.finders.forEach((finder) => {
    config.registerFinder(finder);
  });
  normalizedConfig.actors.forEach((actor) => {
    config.registerActor(actor);
  });
  config.setErrorLevels(normalizedConfig.rules);
  if (configuration.rootElementSelector) {
    config.rootElementSelector = configuration.rootElementSelector;
  }
}
