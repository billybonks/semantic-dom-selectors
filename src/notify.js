import config from './config';

export default function(rule,  type, labelText, generateMessage){
  let rules = config.rules;
  let level = rules[rule];
  if(isNaN(level)){
    level = 0
  }
  let humanizedType = `${type.charAt(0).toUpperCase() + type.slice(1)} Control`
  let message = generateMessage ?
    generateMessage(humanizedType, labelText) :
    buildMessage(rule, humanizedType, labelText);

  switch(level){
    case 0:
      throw new Error(message)
    case 1:
      //eslint-disable-next-line no-console
      console.warn(message)
    }
}


function buildMessage(rule, type, labelText){
  switch(rule){
    case 'ambiguousLabel' :
      return  `Multiple ${type} labelled ${labelText} where found`;
    case 'missingObject' :
      return `Could not find ${type} labelled '${labelText}'`;
    default :
      return `Custom rule ${rule} found ${type} labelled '${labelText}'`
  }
}
