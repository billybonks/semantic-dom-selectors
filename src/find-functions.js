import {
  buttonQuery, textQuery, toggleQuery, selectQuery, formControlQuery,
} from './definitions/selectors';
import config from './config';
import notify from './notify';

const queryHash = {
  text: textQuery,
  toggle: toggleQuery,
  select: selectQuery,
  form: formControlQuery,
  button: buttonQuery,
};

export function findObjects(selector, labelText, type = 'object', index = 0) {
  const finders = config.finders;
  if (finders.length === index) {
    return;
  }
  const finder = finders[index];
  const key = finder.key;
  const strategy = finder.run;

  let objects = strategy(selector, labelText);
  if (!objects || objects.length === 0) {
    objects = findObjects(selector, labelText, type, index + 1);
    if (index === finders.length - 1) {
      return;
    }
  } else if (key !== 'not-aria-compliant') {
    notify(key, type, labelText, finder.errorText);
  }
  return objects || [];
}

export function findObject(selector, labelText, type) {
  const objects = findObjects(selector, labelText, type);
  if (objects.length > 1) {
    notify('ambiguousLabel', type, labelText);
  }
  return objects[0];
}

const _findControl = function (method, labelText, type = 'form') {
  return method(queryHash[type], labelText, type);
};

export function findControl(labelText, type) {
  if (type === 'button') {
    return this.findButton(labelText);
  }
  return _findControl(findObject, labelText, type);
}

export function findControls(labelText, type) {
  return _findControl(findObjects, labelText, type);
}

export function findButton(labelText) {
  return findControl(labelText, 'button');
}

export function findButtons(labelText) {
  return findControls(labelText, 'button');
}
