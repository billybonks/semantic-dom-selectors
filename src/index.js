import { buttonQuery, textQuery, toggleQuery, selectQuery, formControlQuery} from './definitions/selectors';
import config from './config';
import notify from './notify';

let queryHash = {
  text: textQuery,
  toggle: toggleQuery,
  select: selectQuery,
  form: formControlQuery,
  button: buttonQuery,
}


let _findControl = function (method, labelText, type = 'form'){
  return method(queryHash[type], labelText, type)
}

export function findButton(labelText){
  return findControl(labelText, 'button');
}

export function findButtons(labelText){
  return findControls(labelText, 'button');
}

export function findControl(labelText, type) {
  return _findControl(findObject, labelText, type)
}

export default function findControls(labelText, type) {
  return _findControl(findObjects, labelText, type)
}

export function findObject(selector, labelText, type) {
  let objects = findObjects(selector, labelText, type)
  if(objects.length > 1){
    notify('ambiguousLabel', type, labelText);
  }
  return objects[0];
}

export function findObjects(selector, labelText, type='object', index=0) {
  let finders = config.finders;
  if(finders.length === index) {
    return
  }
  let finder = finders[index]
  let key = finder.key
  let strategy = finder.run

  let objects = strategy(selector, labelText)
  if(!objects || objects.length == 0){
    objects = findObjects(selector, labelText, type, index + 1)
    if(index == finders.length-1){
      return
    }
  } else if (key !== 'ariaNotFound') {
    notify(key, type, labelText, finder.errorText);
  }
  return objects || [];
}
