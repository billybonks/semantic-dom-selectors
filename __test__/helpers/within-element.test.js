import { findControl } from '../../src/find-functions';
import withinElement from '../../src/helpers/within-element';
import configure from '../../src/configure';

describe('within-element', () => {
  beforeEach(() => {
    /*
      TODO:invalid-for-relationship-bug
      currently throwing error Control Label of control found through invalid label for relationship
      when it is valid, the bug is in the way notifier works
    */
    configure({ preset: 'default', rules: { 'invalid-label-for': 0 } });
    document.body.innerHTML = `
      <div id="section-one">
        <label for="control">this is my input</label>
        <input id="control" type="text" name="text" />
      </div>
      <div id="section-two">
        <label for="control">this is my input</label>
        <input id="control" type="text" name="text" />
      </div>
    `;
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('if mutliple elements scopes to within section', () => {
    withinElement(document.querySelector('#section-one'), () => {
      const foundInput = findControl('this is my input');
      expect(foundInput).toEqual(document.querySelector('input[id="control"]'));
    });
  });

  test('if no within throws error', () => {
    expect(() => {
      findControl('this is my input');
    }).toThrowErrorMatchingSnapshot();
  });
});
