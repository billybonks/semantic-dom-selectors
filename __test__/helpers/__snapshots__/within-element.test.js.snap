// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within-element if no within throws error 1`] = `"Control this is my input found through invalid label for relationship"`;
